File: backend/app/schemas/self_assessment_summary.py
from datetime import datetime
from typing import List, Optional, Dict, Any
from uuid import UUID
from pydantic import BaseModel, Field


class BucketBreakdown(BaseModel):
    bucket: str
    weight: float
    avg_score: float = Field(..., alias="avgScore")
    contribution: float

    class Config:
        populate_by_name = True


class StageWeights(BaseModel):
    quantitative: float = 0
    qualitative: float = 0
    competency: float = 0


class SelfAssessmentDraftEntry(BaseModel):
    goal_id: UUID = Field(..., alias="goalId")
    bucket: str
    rating_code: Optional[str] = Field(None, alias="ratingCode")
    comment: Optional[str] = None

    class Config:
        populate_by_name = True


class SelfAssessmentContext(BaseModel):
    goals: List[Dict[str, Any]]
    draft: Optional[List[SelfAssessmentDraftEntry]] = None
    stage_weights: StageWeights = Field(..., alias="stageWeights")
    thresholds: List[Dict[str, Any]]
    summary: Optional[Dict[str, Any]] = None

    class Config:
        populate_by_name = True


class SelfAssessmentSummary(BaseModel):
    stage_weights: StageWeights = Field(..., alias="stageWeights")
    per_bucket: List[BucketBreakdown] = Field(..., alias="perBucket")
    weighted_total: float = Field(..., alias="weightedTotal")
    final_rating: str = Field(..., alias="finalRating")
    flags: Dict[str, Any]
    level_adjustment_preview: Optional[Dict[str, Any]] = Field(None, alias="levelAdjustmentPreview")
    submitted_at: Optional[datetime] = Field(None, alias="submittedAt")

    class Config:
        populate_by_name = True

File: frontend/src/api/types/self-assessment-summary.ts
import type { UUID } from './common';

export interface StageWeights {
  quantitative: number;
  qualitative: number;
  competency: number;
}

export interface SelfAssessmentDraftEntry {
  goalId: UUID;
  bucket: string;
  ratingCode?: string;
  comment?: string | null;
}

export interface SelfAssessmentGoalContext {
  id: UUID;
  goalCategory: string;
  periodId: UUID;
  status: string;
  weight?: number | null;
  targetData?: unknown;
}

export interface SelfAssessmentSummaryBucket {
  bucket: string;
  weight: number;
  avgScore: number;
  contribution: number;
}

export interface SelfAssessmentSummaryFlags {
  [key: string]: unknown;
  fail?: boolean;
}

export interface LevelAdjustmentPreview {
  rating?: string;
  delta?: number;
}

export interface SelfAssessmentSummary {
  stageWeights: StageWeights;
  perBucket: SelfAssessmentSummaryBucket[];
  weightedTotal: number;
  finalRating: string;
  flags: SelfAssessmentSummaryFlags;
  levelAdjustmentPreview?: LevelAdjustmentPreview | null;
  submittedAt?: string;
}

export interface ThresholdRow {
  ratingCode: string;
  minScore: number;
  note?: string | null;
}

export interface SelfAssessmentContext {
  goals: SelfAssessmentGoalContext[];
  draft?: SelfAssessmentDraftEntry[] | null;
  stageWeights: StageWeights;
  thresholds: ThresholdRow[];
  summary?: SelfAssessmentSummary | null;
}
